File: treinopro_app/android/app/src/main/kotlin/com/treinopro/treinopro_app/StripeConnectAppearanceFactory.kt
package com.treinopro.app

import android.content.res.Configuration
import android.graphics.Color
import android.os.LocaleList
import com.stripe.android.connect.appearance.Appearance
import com.stripe.android.connect.appearance.Button
import com.stripe.android.connect.appearance.Colors
import com.stripe.android.connect.appearance.CornerRadius
import com.stripe.android.connect.appearance.Form
import org.json.JSONObject
import java.util.Locale

object StripeConnectAppearanceFactory {
    fun buildAppearance(raw: Map<*, *>?): Appearance {
        val appearance = raw ?: emptyMap<Any?, Any?>()
        val primary = parseColor(appearance["primaryColor"] as? String, 0xFFFF6A00.toInt())
        val background = parseColor(appearance["backgroundColor"] as? String, 0xFFFCFDFE.toInt())
        val text = parseColor(appearance["textColor"] as? String, 0xFF0F131A.toInt())
        val secondaryText =
            parseColor(appearance["secondaryTextColor"] as? String, 0xFF616161)
        val danger = parseColor(appearance["dangerColor"] as? String, 0xFFE53D00.toInt())
        val border = parseColor(appearance["borderColor"] as? String, 0xFFE2E8F0.toInt())
        val formBackground =
            parseColor(appearance["formBackgroundColor"] as? String, 0xFFF3F3F3.toInt())
        val formAccent =
            parseColor(appearance["formAccentColor"] as? String, primary)
        val borderRadius = (appearance["borderRadius"] as? Number)?.toFloat() ?: 12f
        val spacingUnit = (appearance["spacingUnit"] as? Number)?.toFloat() ?: 8f

        return Appearance.Builder()
            .colors(
                Colors.Builder()
                    .primary(primary)
                    .background(background)
                    .text(text)
                    .secondaryText(secondaryText)
                    .danger(danger)
                    .border(border)
                    .offsetBackground(background)
                    .build()
            )
            .cornerRadius(
                CornerRadius.Builder()
                    .base(borderRadius)
                    .button(borderRadius)
                    .form(borderRadius)
                    .overlay(borderRadius)
                    .badge(borderRadius)
                    .build()
            )
            .spacingUnit(spacingUnit)
            .buttonPrimary(
                Button(
                    colorBackground = primary,
                    colorBorder = primary,
                    colorText = Color.WHITE,
                )
            )
            .buttonSecondary(
                Button(
                    colorBackground = formBackground,
                    colorBorder = border,
                    colorText = text,
                )
            )
            .form(
                Form.Builder()
                    .colorBackground(formBackground)
                    .accent(formAccent)
                    .build()
            )
            .build()
    }

    fun mergeAppearance(
        local: Map<*, *>?,
        remote: JSONObject?,
    ): Map<String, Any?> {
        val merged = mutableMapOf<String, Any?>()
        local?.forEach { (key, value) ->
            if (key is String && value != null) {
                merged[key] = value
            }
        }

        remote?.optJSONObject("appearance")?.let { remoteAppearance ->
            remoteAppearance.keys().forEach { key ->
                merged[key] = remoteAppearance.get(key)
            }
        }

        remote?.optJSONObject("onboarding")?.let { onboarding ->
            onboarding.optString("locale").takeIf { it.isNotBlank() }?.let {
                merged["locale"] = it
            }
            onboarding.optString("title").takeIf { it.isNotBlank() }?.let {
                merged["title"] = it
            }
            onboarding.optString("privacyPolicyUrl").takeIf { it.isNotBlank() }?.let {
                merged["privacyPolicyUrl"] = it
            }
            onboarding.optString("termsOfServiceUrl").takeIf { it.isNotBlank() }?.let {
                merged["termsOfServiceUrl"] = it
            }
        }

        return merged
    }

    fun applyLocale(activity: MainActivity, localeTag: String) {
        val locale = Locale.forLanguageTag(localeTag.replace('_', '-'))
        Locale.setDefault(locale)
        val config = Configuration(activity.resources.configuration)
        config.setLocales(LocaleList(locale))
        @Suppress("DEPRECATION")
        activity.resources.updateConfiguration(config, activity.resources.displayMetrics)
    }

    private fun parseColor(value: String?, fallback: Int): Int {
        if (value.isNullOrBlank()) {
            return fallback
        }

        return try {
            Color.parseColor(value)
        } catch (_: IllegalArgumentException) {
            fallback
        }
    }
}

File: treinopro_app/ios/Runner/StripeConnectAppearanceFactory.swift
import UIKit
import StripeConnect

enum StripeConnectAppearanceFactory {
  static func buildAppearance(from raw: [String: Any]?) -> EmbeddedComponentManager.Appearance {
    var appearance = EmbeddedComponentManager.Appearance()

    appearance.colors.primary = color(from: raw?["primaryColor"], fallback: 0xFF6A00)
    appearance.colors.background = color(from: raw?["backgroundColor"], fallback: 0xFCFDFE)
    appearance.colors.text = color(from: raw?["textColor"], fallback: 0x0F131A)
    appearance.colors.secondaryText = color(from: raw?["secondaryTextColor"], fallback: 0x616161)
    appearance.colors.danger = color(from: raw?["dangerColor"], fallback: 0xE53D00)
    appearance.colors.border = color(from: raw?["borderColor"], fallback: 0xE2E8F0)
    appearance.colors.formBackground = color(from: raw?["formBackgroundColor"], fallback: 0xF3F3F3)
    appearance.colors.formAccent = color(from: raw?["formAccentColor"], fallback: 0xFF6A00)
    appearance.colors.actionPrimaryText = appearance.colors.primary

    let borderRadius = number(from: raw?["borderRadius"], fallback: 12)
    appearance.cornerRadius.base = borderRadius
    appearance.cornerRadius.button = borderRadius
    appearance.cornerRadius.form = borderRadius
    appearance.cornerRadius.overlay = borderRadius
    appearance.cornerRadius.badge = borderRadius
    appearance.spacingUnit = number(from: raw?["spacingUnit"], fallback: 8)

    appearance.buttonPrimary.colorBackground = appearance.colors.primary
    appearance.buttonPrimary.colorBorder = appearance.colors.primary
    appearance.buttonPrimary.colorText = .white

    appearance.buttonSecondary.colorBackground = appearance.colors.formBackground
    appearance.buttonSecondary.colorBorder = appearance.colors.border
    appearance.buttonSecondary.colorText = appearance.colors.text

    return appearance
  }

  static func mergeAppearance(
    local: [String: Any]?,
    remote: [String: Any]?
  ) -> [String: Any] {
    var merged = local ?? [:]

    if let remoteAppearance = remote?["appearance"] as? [String: Any] {
      for (key, value) in remoteAppearance {
        merged[key] = value
      }
    }

    if let onboarding = remote?["onboarding"] as? [String: Any] {
      if let locale = onboarding["locale"] as? String, !locale.isEmpty {
        merged["locale"] = locale
      }
      if let title = onboarding["title"] as? String, !title.isEmpty {
        merged["title"] = title
      }
      if let privacyPolicyUrl = onboarding["privacyPolicyUrl"] as? String, !privacyPolicyUrl.isEmpty {
        merged["privacyPolicyUrl"] = privacyPolicyUrl
      }
      if let termsOfServiceUrl = onboarding["termsOfServiceUrl"] as? String, !termsOfServiceUrl.isEmpty {
        merged["termsOfServiceUrl"] = termsOfServiceUrl
      }
    }

    return merged
  }

  private static func color(from value: Any?, fallback: UInt32) -> UIColor {
    guard let hex = value as? String else {
      return uiColor(from: fallback)
    }

    var sanitized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
    if sanitized.hasPrefix("#") {
      sanitized.removeFirst()
    }

    guard sanitized.count == 6, let intValue = UInt32(sanitized, radix: 16) else {
      return uiColor(from: fallback)
    }

    return uiColor(from: intValue)
  }

  private static func uiColor(from hex: UInt32) -> UIColor {
    UIColor(
      red: CGFloat((hex >> 16) & 0xFF) / 255.0,
      green: CGFloat((hex >> 8) & 0xFF) / 255.0,
      blue: CGFloat(hex & 0xFF) / 255.0,
      alpha: 1.0
    )
  }

  private static func number(from value: Any?, fallback: CGFloat) -> CGFloat {
    if let number = value as? NSNumber {
      return CGFloat(truncating: number)
    }
    if let doubleValue = value as? Double {
      return CGFloat(doubleValue)
    }
    return fallback
  }
}
